File: bindings/ios/Sources/Collection.swift
import Foundation

/// A vector collection (table).
/// Created via ``VecDB/createCollection(name:dimension:)``.
/// Must call ``close()`` when done.
public final class Collection {
    private let ptr: OpaquePointer?
    private var closed = false

    init(db: OpaquePointer?, name: String, dimension: Int) {
        var cName = name.cString(using: .utf8)!
        ptr = jw_collection_create(db, &cName, UInt32(dimension))
        if ptr == nil {
            fatalError("jinwo_vecdb: failed to create collection '\(name)'")
        }
    }

    /// Init from an existing collection pointer (for getCollection / reopen).
    init(existing: OpaquePointer?, name: String) {
        self.ptr = existing
    }

    deinit { close() }

    /// Insert a vector. Returns the assigned vector ID.
    @discardableResult
    public func insert(_ vector: [Float]) throws -> UInt64 {
        guard !closed else { throw VecDBError.collectionNotFound }
        let dim = dimension()
        guard vector.count == dim else {
            throw VecDBError.invalidDimension
        }
        let rc = vector.withUnsafeBufferPointer { buf in
            jw_collection_insert(ptr, buf.baseAddress, UInt32(dim))
        }
        if rc != 0 { throw VecDBError(rawValue: Int(rc)) ?? .unknown }
        return 0 // last insert ID approximation
    }

    /// Insert a vector with correct VID return.
    @discardableResult
    public func insertVid(_ vector: [Float]) throws -> UInt64 {
        guard !closed else { throw VecDBError.collectionNotFound }
        let dim = dimension()
        guard vector.count == dim else {
            throw VecDBError.invalidDimension
        }
        let vid = vector.withUnsafeBufferPointer { buf in
            jw_collection_insert_vid(ptr, buf.baseAddress, UInt32(dim))
        }
        if vid == 0 { throw VecDBError.unknown }
        return vid
    }

    /// Get a vector by ID.
    public func get(vid: UInt64) throws -> [Float] {
        guard !closed else { throw VecDBError.collectionNotFound }
        let dim = dimension()
        var vec = [Float](repeating: 0, count: dim)
        let rc = vec.withUnsafeMutableBufferPointer { buf in
            jw_collection_get_vector(ptr, vid, buf.baseAddress)
        }
        if rc != 0 { throw VecDBError.vectorNotFound }
        return vec
    }

    /// Count of vectors in the collection.
    public func count() -> Int {
        guard !closed else { return 0 }
        return Int(jw_collection_count(ptr))
    }

    /// Delete a vector by ID.
    public func delete(vid: UInt64) throws {
        guard !closed else { throw VecDBError.collectionNotFound }
        let rc = jw_collection_delete(ptr, vid)
        if rc != 0 { throw VecDBError(rawValue: Int(rc)) ?? .unknown }
    }

    /// Search for k nearest neighbours.
    public func search(query: [Float], k: Int) throws -> [SearchResult] {
        guard !closed else { throw VecDBError.collectionNotFound }
        let dim = dimension()
        guard query.count == dim else {
            throw VecDBError.invalidDimension
        }
        var results = [jw_search_result](repeating: jw_search_result(id: 0, distance: 0), count: k)
        let count = query.withUnsafeBufferPointer { buf in
            jw_collection_search(ptr, buf.baseAddress, UInt32(dim), Int32(k), &results)
        }
        if count < 0 { throw VecDBError.indexNotReady }
        return Array(results.prefix(Int(count))).map {
            SearchResult(id: $0.id, distance: $0.distance)
        }
    }

    /// Build search index (required before searching).
    public func buildIndex() throws {
        guard !closed else { throw VecDBError.collectionNotFound }
        let rc = jw_collection_build_index(ptr)
        if rc != 0 { throw VecDBError(rawValue: Int(rc)) ?? .unknown }
    }

    /// Encoding dimension of the collection.
    public func dimension() -> Int {
        Int(jw_collection_encoding_dim(ptr))
    }

    /// Close the collection.
    /// Does NOT destroy the underlying collection — the DB owns it
    /// and will clean up all collections when the DB is closed.
    public func close() {
        closed = true
    }
}

File: bindings/android/library/src/main/java/com/jinwo/vecdb/Collection.java
package com.jinwo.vecdb;

import java.util.ArrayList;
import java.util.List;

/**
 * VecDB collection (table of vectors).
 * <p>
 * Create via {@link JinWoDB#createCollection(String, int)}.
 * Must call {@link #close()} when done.
 */
public class Collection implements AutoCloseable {

    private final long ptr;
    private boolean closed = false;

    /** JNI: create collection */
    private static native long nativeCreateCollection(long dbPtr, String name, int dimension);

    /** JNI: close collection */
    private static native void nativeCloseCollection(long ptr);

    /** JNI: insert vector, returns vector ID */
    private static native long nativeInsert(long ptr, float[] vector, int dim);

    /** JNI: delete vector by ID */
    private static native int nativeDelete(long ptr, long vid);

    /** JNI: search KNN */
    private static native long[] nativeSearch(long ptr, float[] query, int dim, int k);

    /** JNI: build index */
    private static native int nativeBuildIndex(long ptr);

    /** JNI: get encoding dimension */
    private static native int nativeGetDimension(long ptr);

    /** JNI: get vector by ID */
    private static native float[] nativeGetVector(long ptr, long vid);

    /** JNI: get collection count */
    private static native long nativeGetCount(long ptr);

    Collection(long dbPtr, String name, int dimension) {
        this.ptr = nativeCreateCollection(dbPtr, name, dimension);
        if (this.ptr == 0) {
            throw new RuntimeException("jinwo_vecdb: failed to create collection '" + name + "'");
        }
    }

    /** Construct from existing collection pointer (for getCollection / reopen). */
    Collection(long ptr) {
        this.ptr = ptr;
    }

    long getPtr() { return ptr; }

    /**
     * Insert a vector.
     * @return the assigned vector ID.
     */
    public long insert(float[] vector) {
        checkClosed();
        int dim = getDimension();
        if (vector.length != dim) {
            throw new IllegalArgumentException("jinwo_vecdb: dimension mismatch: expected " + dim + ", got " + vector.length);
        }
        return nativeInsert(ptr, vector, dim);
    }

    /**
     * Delete a vector by its ID.
     */
    public void delete(long vid) {
        checkClosed();
        int rc = nativeDelete(ptr, vid);
        if (rc != 0) {
            throw new RuntimeException("jinwo_vecdb: delete failed (code=" + rc + ")");
        }
    }

    /**
     * Search for the k nearest neighbours.
     */
    public List<SearchResult> search(float[] query, int k) {
        checkClosed();
        int dim = getDimension();
        if (query.length != dim) {
            throw new IllegalArgumentException("jinwo_vecdb: dimension mismatch: expected " + dim + ", got " + query.length);
        }
        long[] raw = nativeSearch(ptr, query, dim, k);
        if (raw == null) return new ArrayList<>();

        List<SearchResult> results = new ArrayList<>(raw.length / 2);
        for (int i = 0; i < raw.length; i += 2) {
            long id = raw[i];
            float dist = Float.intBitsToFloat((int) raw[i + 1]);
            results.add(new SearchResult(id, dist));
        }
        return results;
    }

    /**
     * Build search index (required before searching).
     */
    public void buildIndex() {
        checkClosed();
        int rc = nativeBuildIndex(ptr);
        if (rc != 0) {
            throw new RuntimeException("jinwo_vecdb: buildIndex failed (code=" + rc + ")");
        }
    }

    /**
     * Get a vector by its ID.
     */
    public float[] get(long vid) {
        checkClosed();
        float[] vec = nativeGetVector(ptr, vid);
        if (vec == null) {
            throw new RuntimeException("jinwo_vecdb: get failed for vid=" + vid);
        }
        return vec;
    }

    /**
     * Get number of vectors in the collection.
     */
    public long count() {
        checkClosed();
        return nativeGetCount(ptr);
    }

    /**
     * Get the encoding dimension of this collection.
     */
    public int getDimension() {
        checkClosed();
        return nativeGetDimension(ptr);
    }

    @Override
    public void close() {
        if (!closed) {
            nativeCloseCollection(ptr);
            closed = true;
        }
    }

    private void checkClosed() {
        if (closed) throw new IllegalStateException("Collection is closed");
    }
}
